modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C

```diff
@@ -238,85 +355,150 @@ TemperaturePressureFunctionFluidProperties::h_from_p_T(
 }
 
 Real
-TemperaturePressureFunctionFluidProperties::e_from_p_T(Real /* pressure */, Real temperature) const
-{
-  return _cv * temperature;
+TemperaturePressureFunctionFluidProperties::e_from_p_T(Real pressure, Real temperature) const
+{
+  if (_cv_is_constant)
+    return _e_ref + _cv * (temperature - _T_ref);
+  else
+  {
+    // Use Simpson's 3/8 rule to integrate cv
+    Real h = (temperature - _T_ref) / 3;
+    Real integral =
+        3. / 8. * h *
+        (cv_from_p_T(pressure, _T_ref) + 3 * cv_from_p_T(pressure, _T_ref + h) +
+         3 * cv_from_p_T(pressure, _T_ref + 2 * h) + cv_from_p_T(pressure, temperature));
+    return _e_ref + integral;
```

Comment:
I think you suggested this in another comment, but maybe a parameter with a numerical integral interval would be best.

Reply:
done

Reply:
What are your thoughts on an external Simpson's integration method (my second comment)?

Reply:
Ok, not a required change.